src/electos/ballotmaker/data/extractor.py

```diff
@@ -0,0 +1,311 @@
+        return district
+
+
+    def _candidate_contest_of(self, contest: CandidateContest):
+        """Extract candidate contest subset needed for a ballot."""
+        district = self._contest_election_district(contest)
+        offices = self._candidate_contest_offices(contest)
+        parties = self._candidate_contest_parties(contest)
+        candidates = self._candidate_contest_candidates(contest)
+        result = {
+            "id": contest.model__id,
+            "type": "candidate",
+            "title": contest.name,
+            "district": district,
+            "vote_type": contest.vote_variation.value,
+            # Include even when default is 1: don't require caller to track that.
+            "votes_allowed": contest.votes_allowed,
+            "candidates": candidates,
+            # "offices": offices,
+            # "parties": parties,
+        }
+        return result
+
+
+    def _ballot_measure_contest_of(self, contest: BallotMeasureContest):
+        """Extract ballot measure contest subset needed for a ballot."""
+        choices = []
+        for selection in contest.contest_selection:
+            assert isinstance(selection, BallotMeasureSelection), \
+               f"Unexpected non-ballot measure selection: {type(selection).__name__}"
+            choice = _text_content(selection.selection)
+            choices.append({
+                "id": selection.model__id,
+                "choice": choice,
+            })
+        district = self._contest_election_district(contest)
+        full_text = _text_content(contest.full_text)
+        result = {
+            "id": contest.model__id,
+            "type": "ballot measure",
+            "title": contest.name,
+            "district": district,
+            "text": full_text,
+            "choices": choices,
+        }
+        return result
+
+
+    def _contests(self, ballot_style: BallotStyle):
+        """Extract contest subset needed for ballots."""
+        for contest in self._ballot_style_contests(ballot_style):
+            if isinstance(contest, CandidateContest):
+                entry = self._candidate_contest_of(contest)
+            elif isinstance(contest, BallotMeasureContest):
+                entry = self._ballot_measure_contest_of(contest)
+            else:
+                # Ignore other contest types
+                print(f"Skipping contest of type {contest.model__type}")
+            yield entry
+
+
+    def _election_ballot_styles(self, election: Election):
+        """Extract all ballot styles."""
+        for ballot_style in election.ballot_style:
+            data = {
+                "id": self._ballot_style_external_id(ballot_style),
+                "scopes": [
+                    _.model__id
+                    for _ in self._ballot_style_gp_units(ballot_style)
+                ],
+                "contests": [
+                    _
+                    for _ in self._contests(ballot_style)
```

Comment:
It's really supposed to be this:

```
"contests: [  _  for _ in self._contests(ballot_style) ],
```

The reason it's not is that the nesting was driving the line length past the point where `black` will wrap it anyway. I probably should have just left it long. 

This a convention I've seen in comprehensions that indicates that all that matters is what's done with the result, and keeps them short (one-liners). I don't like one letter variables, don't have a good alternative, and `_` has a meaning in Python: variable that's being ignored. If you have another sensible short identifier to suggest I can use that. (I could use `it` I suppose.)

But I don't want to be returning lists. This is a performance issue: Callees yield instead of returning lists because generators use constant memory, and lists use O(N) memory. Callees can make lists if they need them. This isn't going to matter for one demo PDF. It may not matter for lots of them. But as a general technique this means arbitrary objects are not being created unnecessarily, and that can matter when things scale up.